### PR TITLE
PRO-2501-Arka_Contract_Events_Whitelist

### PR DESCRIPTION
**Pull request by vignesha22**
## Description

- Added a new table for Contract whitelisting based on the sponsor wallet address
- Added new env var for EPv07 token paymasters VGL and PGL
- Added migrations as modicifactions are done on apiKeys to include another column for enabling contract whitelisting

## Types of changes
What types of changes does your code introduce?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Further comments (optional)

-

### CODE REVIEW

File: backend/migrations/2024080800001-create-contract-whitelist.cjs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    },
+    WALLET_ADDRESS: {
+      type: Sequelize.TEXT,
+      allowNull: false,
+    },
+    CONTRACT_ADDRESS: {
+      type: Sequelize.TEXT,
+      allowNull: false,
+    },
+    EVENT_NAMES: {
```
Comment (ch4r10t33r) [inline backend/migrations/2024080800001-create-contract-whitelist.cjs:19]: Call it FUNCTION_SELECTOR. Events are emitted within a contract.

File: backend/src/models/contract-whitelist.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      type: DataTypes.TEXT,
+      allowNull: false,
+      field: 'WALLET_ADDRESS'
+    },
+    contractAddress: {
+      type: DataTypes.TEXT,
+      allowNull: false,
+      field: 'CONTRACT_ADDRESS'
+    },
+    eventNames: {
```
Comment (ch4r10t33r) [inline backend/src/models/contract-whitelist.ts:32]: Pls see above, rename this to functionSelector.

File: backend/src/routes/whitelist-routes.ts [hunk 3]
```diff
@@ -485,6 +478,199 @@ const whitelistRoutes: FastifyPluginAsync = async (server) => {
       }
     }
   )
+
+  server.post("/contractAddWhitelist",
```
Comment (ch4r10t33r) [inline backend/src/routes/whitelist-routes.ts:482]: can we rename this as `whiteListContractAddress`?

File: backend/src/routes/whitelist-routes.ts [hunk 4]
```diff
@@ -485,6 +478,199 @@ const whitelistRoutes: FastifyPluginAsync = async (server) => {
...
+      } catch (err: any) {
+        request.log.error(err);
+        if (err.name == "ResourceNotFoundException")
+          return reply.code(ReturnCode.FAILURE).send({ error: ErrorMessage.INVALID_API_KEY });
+        return reply.code(ReturnCode.FAILURE).send({ error: err.message ?? ErrorMessage.FAILED_TO_CREATE_CONTRACT_WHITELIST });
+      }
+    }
+  )
+
+  server.post("/contractUpdateWhitelist",
```
Comment (ch4r10t33r) [inline backend/src/routes/whitelist-routes.ts:547]: rename to `updateWhitelistContractAddress`

File: backend/src/routes/whitelist-routes.ts [hunk 5]
```diff
@@ -485,6 +478,199 @@ const whitelistRoutes: FastifyPluginAsync = async (server) => {
...
+        request.log.error(err);
+        if (err.name == "ResourceNotFoundException")
+          return reply.code(ReturnCode.FAILURE).send({ error: ErrorMessage.INVALID_API_KEY });
+        return reply.code(ReturnCode.FAILURE).send({ error: err.message ?? ErrorMessage.FAILED_TO_UPDATE_CONTRACT_WHITELIST });
+      }
+    }
+  )
+
+
+  server.post("/contractDeleteWhitelist",
```
Comment (ch4r10t33r) [inline backend/src/routes/whitelist-routes.ts:613]: rename to `deleteWhitelistContractAddress`

File: backend/src/types/contractWhitelist-dto.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+
+export interface ContractWhitelistDto {
+  walletAddress: string;
+  contractAddress: string;
+  eventNames: string[];
```
Comment (ch4r10t33r) [inline backend/src/types/contractWhitelist-dto.ts:5]: must be called `functionSelector`

**Review by ch4r10t33r — CHANGES_REQUESTED** [review]
Have we shortlisted the list of contract addresses which we intend to preload to the hosted setup? Will this be included in this PR?

**Review by ch4r10t33r — APPROVED** [review]
LGTM

**Review by kanthgithub — APPROVED** [review]

**Pull request merged**